linkerd/app/outbound/src/http/breaker/consecutive_failures.rs

```diff
@@ -0,0 +1,212 @@
+    gate: gate::Tx,
+    rsps: mpsc::Receiver<classify::Class>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl ConsecutiveFailures {
+    pub fn new(
+        max_failures: usize,
+        backoff: ExponentialBackoff,
+        gate: gate::Tx,
+        rsps: mpsc::Receiver<classify::Class>,
+    ) -> Self {
+        Self {
+            max_failures,
+            backoff,
+            gate,
+            rsps,
+            semaphore: Arc::new(Semaphore::new(0)),
+        }
+    }
+
+    pub(super) async fn run(mut self) {
+        loop {
+            if self.open().await.is_err() {
+                return;
+            }
+
+            if self.closed().await.is_err() {
+                return;
+            }
+        }
+    }
+
+    /// Keep the breaker open until `max_failures` consecutive failures are
+    /// observed.
+    async fn open(&mut self) -> Result<(), ()> {
+        tracing::debug!("Open");
+        self.gate.open();
+        let mut failures = 0;
+        loop {
+            let class = tokio::select! {
+                rsp = self.rsps.recv() => rsp.ok_or(())?,
+                _ = self.gate.lost() => return Err(()),
+            };
+
+            tracing::trace!(?class, %failures, "Response");
+            if class.is_success() {
+                failures = 0;
+            } else {
+                failures += 1;
+                if failures == self.max_failures {
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    /// Keep the breaker closed for at least the initial backoff, and then,
+    /// once the timeout expires, go into probation to admit a single request
+    /// before reverting to the open state or continuing in the shut state.
+    async fn closed(&mut self) -> Result<(), ()> {
+        let mut backoff = self.backoff.stream();
+        loop {
+            // The breaker is shut now. Wait until we can open it again.
+            tracing::debug!(backoff = ?backoff.duration(), "Shut");
+            self.gate.shut();
+
+            loop {
+                tokio::select! {
+                    _ = backoff.next() => break,
+                    // Ignore responses while the breaker is shut.
+                    _ = self.rsps.recv() => continue,
+                    _ = self.gate.lost() => return Err(()),
+                }
+            }
+
+            let class = self.probation().await?;
+            tracing::trace!(?class, "Response");
+            if class.is_success() {
+                // Open!
+                return Ok(());
+            }
+        }
+    }
+
+    /// Wait for a response to determine whether the breaker should be opened.
+    async fn probation(&mut self) -> Result<classify::Class, ()> {
+        tracing::debug!("Probation");
+        self.semaphore.add_permits(1);
+        self.gate.limit(self.semaphore.clone());
+        tokio::select! {
+            rsp = self.rsps.recv() => rsp.ok_or(()),
+            _ = self.gate.lost() => Err(()),
+        }
```

Comment:
I don't think this is a blocker (and I'm not really sure how to resolve it off the top of my head), but the current classifiers are going to behave poorly with streaming responses:

If we're in probation and the first request we allow receives a streaming response, we won't process the response summary until the end of the stream! So if that stream stays open for 10 minutes, we won't exit probation for 10 minutes. And if the streaming response ends with an error code, then we'll go back into shut after a 10 minute delay!

Reply:
I suppose we could impose a max probation time limit: if we don't exit within that period, just open the gate.

Reply:
@hawkw It seems reasonable to use the max backoff value as the time limit...

Reply:
Yeah, that makes sense --- I'll add a probation limit. Would you prefer it in this PR or a follow-up?